src/passes/Unsubtyping.cpp

```diff
@@ -0,0 +1,580 @@
+// types are related, we do need to make sure that we do not change the
+// behavior of casts by removing subtype relationships they might observe. For
+// example, consider this module:
+//
+// (module
+//  ;; original subtyping: $bot <: $mid <: $top
+//  (type $top (sub (struct)))
+//  (type $mid (sub $top (struct)))
+//  (type $bot (sub $mid (struct)))
+//
+//  (func $f
+//   (local $top (ref $top))
+//   (local $mid (ref $mid))
+//
+//   ;; Requires $bot <: $top
+//   (local.set $top (struct.new $bot))
+//
+//   ;; Cast $top to $mid
+//   (local.set $mid (ref.cast (ref $mid) (local.get $top)))
+//  )
+// )
+//
+// The only subtype relation directly required by the IR for this module is $bot
+// <: $top. However, if we optimized the module so that $bot <: $top was the
+// only subtype relation, we would change the behavior of the cast. In the
+// original module, a value of type (ref $bot) is cast to (ref $mid). The cast
+// succeeds because in the original module, $bot <: $mid. If we optimize so that
+// we have $bot <: $top and no other subtypings, though, the cast will fail
+// because the value of type (ref $bot) no longer inhabits (ref $mid). To
+// prevent the cast's behavior from changing, we need to ensure that $bot <:
+// $mid.
+//
+// The set of subtyping requirements generated by a cast from $src to $dest is
+// that for every known remaining subtype $v of $src, if $v <: $dest in the
+// original module, then $v <: $dest in the optimized module. In other words,
+// for every type $v of values we know can flow into the cast, if the cast would
+// have succeeded for values of type $v before, then we know the cast must
+// continue to succeed for values of type $v. These requirements arising from
+// casts can also generate transitive requirements because we learn about new
+// types of values that can flow into casts as we learn about new subtypes of
+// cast sources.
+//
+// Starting with the initial subtype relation determined by walking the IR,
+// repeatedly search for new subtypings by analyzing type definitions and casts
+// in lock step until we reach a fixed point. This is the minimal subtype
+// relation that preserves module validity and behavior that can be found
+// without a more precise analysis of types that might flow into each cast.
+
+namespace wasm {
+
+namespace {
+
+struct Unsubtyping
+  : WalkerPass<ControlFlowWalker<Unsubtyping, OverriddenVisitor<Unsubtyping>>> {
+  // The new set of supertype relations.
+  std::unordered_map<HeapType, HeapType> supertypes;
+
+  // Map from cast source types to their destinations.
+  std::unordered_map<HeapType, std::unordered_set<HeapType>> castTypes;
+
+  // The set of subtypes that need to have their type definitions analyzed to
+  // transitively find other subtype relations they depend on. We add to it
+  // every time we find a new subtype relationship we need to keep.
+  std::unordered_set<HeapType> worklist;
```

Comment:
Might be worth seeing if using `UniqueDeferredQueue` instead of `unordered_set` is faster here. It usually is in my experience, and as a bonus, there is a simple `push/pop` so using the API is more convenient later down.

Reply:
Indeed, this saves a second and a half on the unoptimized binary 👍